[UIST] Add new object type UIST

ViktoriaFreidel opened the pull request.

ViktoriaFreidel committed:
[UIST] Add new object type UIST

ViktoriaFreidel committed:
[UIST] Add new object type UIST

schneidermic0 commented on file-formats/uist/examples/z_aff_example_uist.uist.json:
```diff
@@ -0,0 +1,18 @@
+{
```
It seems to me that this is an issue in our validation for JSON examples. Do I miss anything?

larshp replied:
there is https://github.com/python-jsonschema/jsonschema/issues/1039, but I have no idea, its Python...

schneidermic0 replied:
I have just created an issue #498

schneidermic0 replied:
@ViktoriaFreidel  JFYI. Please ignore the error for now. I will also ask you to remove the multiple of value from your ABAP type

ViktoriaFreidel replied:
Ok. No problem. Today I will send update taday.

schneidermic0 reviewed (changes requested):
Thanks, @ViktoriaFreidel for your pull request.

I reviewed your proposal and added some questions and comments.

wurzka commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,67 @@
+    "! <p class="shorttext">Fiori Launchpad Space Template</p>
+    "! Fiori Launchpad Space Template
+    BEGIN OF ty_main,
+      "! <p class="shorttext">Format version</p>
```
```suggestion
      "! <p class="shorttext">Format Version</p>
```

ViktoriaFreidel commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,67 @@
+      "! Base Space Template ID
+      base_space_template_id       TYPE c LENGTH 35,
+
+      "! <p class="shorttext">Assigned SAP Fiori Launchpad Page Templates</p>
+      "! Assigned SAP Fiori launchpad page templates
+      pages                        TYPE ty_pages,
```
Yes, it is correct. Thanks.

ViktoriaFreidel committed:
Update file-formats/uist/type/zif_aff_uist_v1.intf.abap

Co-authored-by: Michael Schneider <micha.schneider@sap.com>

ViktoriaFreidel committed:
Update file-formats/uist/type/zif_aff_uist_v1.intf.abap

Ok. It maybe is a reason why I got a warning message about a format version in generation report.

Co-authored-by: Katharina Wurz <katharina.wurz@sap.com>

larshp commented:
in JS, it is a IEEE floating point, but displays correctly,

![image](https://user-images.githubusercontent.com/5888506/222440275-392ef551-4fc8-4296-ac53-e19bb4a8d67d.png)

ViktoriaFreidel commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,67 @@
+      "! ID
+      page_template_id          TYPE c LENGTH 35,
+
+      "! <p class="shorttext">Title</p>
+      "! Title
+      page_template_title       TYPE zif_aff_types_v1=>ty_description_100,
+
+      "! <p class="shorttext">Description</p>
+      "! Description
+      page_template_description TYPE zif_aff_types_v1=>ty_description_100,
```
You are right. In space template only assigned Page Id is stored. Title and description of pages  are stored in own pages table. I will update a space template model

ViktoriaFreidel replied:
Question: if I remove page title and page description how can I display on UI Page description and title :
![image](https://user-images.githubusercontent.com/126160451/223681459-6509321d-689d-43df-94d6-4175d000c340.png)

schneidermic0 commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,67 @@
+      "! Sort priority
+      "! $minimum: 0.001
+      "! $maximum: 999.999
+      "! $multipleOf: 0.001
```
Would you remove the annotation `$multipleOf`, please? We face issues with the validation and will update our generator accordingly
```suggestion
      "! $multipleOf: 0.001
```

schneidermic0 commented on file-formats/uist/uist-v1.json:
```diff
@@ -0,0 +1,134 @@
+      "description": "Sort priority",
+      "type": "number",
+      "minimum": 0.001,
+      "maximum": 999.999,
+      "multipleOf": 0.001
```
Most probably, the automatic generation will regenerate the property `"multipleOf"`to your schema. We plan to fix this.

```suggestion
      "maximum": 999.999
```

schneidermic0 replied:
I think we have to remove the `multipleOf` manually, currently. We plan to solve this in the generator soon.

@larshp Or do you think we shall keep it for now (until the generator is fixed) otherwise the validation will raise an issue, won't it?

schneidermic0 reviewed (commented):
I have just added some proposals to get rid of the "multipleOf"-issue we see in this PR

ViktoriaFreidel commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,67 @@
+      "! <p class="shorttext">ID</p>
+      "! ID
+      "! $required
+      space_template_id            TYPE c LENGTH 35,
```
No, this is an Id for space template itself

ViktoriaFreidel commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,67 @@
+      "! $required
+      space_template_id            TYPE c LENGTH 35,
+
+      "! <p class="shorttext">Title</p>
+      "! Title
+      "! $required
+      space_template_title         TYPE zif_aff_types_v1=>ty_description_100,
+
+      "! <p class="shorttext">Description </p>
+      "! Description
+      "! $required
+      space_template_description   TYPE zif_aff_types_v1=>ty_description_100,
```
No, the description, title belongs to space template and header-description will be synced  with description if it changed.

schneidermic0 replied:
I understand you store data redundantly in two fields. We should avoid this. Can we remove one of the fields?

ViktoriaFreidel replied:
I see your point. I try to remove field "space_template_description" (I will rename it ) and set/used from header-description. 
I wasn't  able to set description fields on UI under Title, means in a specific position. '"Header" Fields are on UI on fix position and I can not hidden fields one by one?
![image](https://user-images.githubusercontent.com/126160451/223698811-6160956e-aaa4-4e50-ab3b-f84cd0f301fe.png)
If you could give me some suggestion to solve UI, then we could delete description field.

schneidermic0 replied:
The description of an object is changed in the Properties view in ADT

schneidermic0 replied:
As discussed, let's remove the field `description` on top level. field `header-description` is sufficient.

ViktoriaFreidel commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,67 @@
+      "! <p class="shorttext">Title</p>
+      "! Title
+      "! $required
+      space_template_title         TYPE zif_aff_types_v1=>ty_description_100,
```
OK, I removed it with a last change

ViktoriaFreidel commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,67 @@
+      "! <p class="shorttext">Header</p>
+      "! Header
+      "! $required
+      header                       TYPE zif_aff_types_v1=>ty_header_100,
```
Usage in "ABAP cloud development" - is yes. Definitively
For Usage in "Key User" - I forward this question to PO

schneidermic0 replied:
Sorry, I didn't see you updated the comment. 

I guessed you are not using ABAP language version "Key user". In this case, we should replace the type by another type. I'll clarify with @SAP/abap-file-formats-team whether we introduce `zif_aff_types_v1=>ty_header_100_cloud` or whether we shall create an own type in your interface.

schneidermic0 replied:
I have just talked to the team. We will extend the interface `zif_aff_types_v1` with a new type `ty_header_100_cloud`

ViktoriaFreidel committed:
Add new object type UIST

ViktoriaFreidel committed:
Add new object type UIST

ViktoriaFreidel committed:
[UIST] Add new object type UIST

schneidermic0 committed:
Merge branch 'main' into feature/uist

Commit:
Add header 100 for cloud

schneidermic0 reviewed (changes requested):
I tried to summarise what we discussed earlier today

ViktoriaFreidel committed:
Update /uist/type/zif_aff_uist_v1.intf.abap use cloud description

Co-authored-by: Michael Schneider <micha.schneider@sap.com>

ViktoriaFreidel committed:
Update zif_aff_uist_v1.intf.abap - use Name and not Id

Co-authored-by: Michael Schneider <micha.schneider@sap.com>

ViktoriaFreidel committed:
Update zif_aff_uist_v1.intf.abap - use Page Name not Id

Co-authored-by: Michael Schneider <micha.schneider@sap.com>

ViktoriaFreidel committed:
Update zif_aff_uist_v1.intf.abap- remove pages description and title

Co-authored-by: Michael Schneider <micha.schneider@sap.com>

ViktoriaFreidel committed:
[UIST] Add new object type UIST and Update  AFF Model

ViktoriaFreidel committed:
[UIST] Add new object type UIST and Update  AFF Model

ViktoriaFreidel committed:
[UIST] Add new object type UIST and Update  AFF Model

schneidermic0 reviewed (changes requested):
Thank you for making so many improvements. I have added some small question.

ViktoriaFreidel commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -32,35 +46,15 @@ INTERFACE zif_aff_uist_v1
       "! Header
       "! $required
-      header          TYPE zif_aff_types_v1=>ty_header_100,
+      header          TYPE zif_aff_types_v1=>ty_header_60_cloud,
```
It is my mistake, I generate a files from Interface in ABAP and there ty_header_100_cloud is not available yet. Temporary I set to _60_ to avoid abap error and forgot to change is in a file.

ViktoriaFreidel commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -32,35 +46,15 @@ INTERFACE zif_aff_uist_v1
-      "! <p class="shorttext">Base Space Template ID</p>
-      "! Base Space Template ID
-      base_space_template_id       TYPE c LENGTH 35,
+      properties   TYPE ty_properties,
```
I will rename it.

ViktoriaFreidel committed:
[UIST] Add new object type UIST and Update AFF Model

ViktoriaFreidel committed:
[UIST] Add new object type UIST

schneidermic0 commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,56 @@
+  PUBLIC.
+
+  TYPES:
+    "! <p class="shorttext">Properties</p>
+    "! Properties
+    BEGIN OF ty_general_information,
```
I suggest to adapt also ABAP Doc accordingly

```suggestion
    "! <p class="shorttext">General Information</p>
    "! General information
    BEGIN OF ty_general_information,
```

schneidermic0 commented on file-formats/uist/examples/z_aff_example_uist.uist.json:
```diff
@@ -0,0 +1,18 @@
+  "formatVersion": "1",
+  "header": {
+    "description": "Example of UIST Object for ABAP file formats",
+    "originalLanguage": "en",
+    "abapLanguageVersion": "standard"
```
Since `"abapLanguageVersion": "standard"` is the default, we usually don't render it. Therefore, I would just remove the value from the example

```suggestion
    "originalLanguage": "en"
```

schneidermic0 commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,56 @@
+      "! Assigned SAP Fiori launchpad page templates
+      pages                        TYPE ty_pages,
+    END OF ty_main.
+ENDINTERFACE.
```
To fix the editorconfig issue

```suggestion
ENDINTERFACE.

```

schneidermic0 reviewed (changes requested):
Thanks for the changes. I have found just three minor issue. I will ask my team members to double check it, too.

schneidermic0 commented:
@ViktoriaFreidel Did you also sign the contributor license agreement (CLA)?

wurzka commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,56 @@
+      "! $maximum: 999.999
+      sort_priority TYPE p LENGTH 7 DECIMALS 3,
+      "! <p class="shorttext">Base Space Template Name</p>
+      "! Base Space Template Name
```
I'm not sure about SAP Terms here, but I think at least 'name' should be lower case, no?
```suggestion
      "! Base Space Template name
```

burcuka replied:
maybe we could change it to "Name of the Base Space Template"? also not sure about the writing of "Base Space Template"

ViktoriaFreidel replied:
We change it to "Name of the base space template" - as description .
... Write the description in sentence case .... - from documentation.

wurzka commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,56 @@
+    "! Properties
+    BEGIN OF ty_general_information,
+      "! <p class="shorttext">Title</p>
+      "! Space Template Title
```
Same 
```suggestion
      "! Space Template title
```

wurzka commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,56 @@
+    "! SAP Fiori Page Template
+    BEGIN OF ty_page,
+      "! <p class="shorttext">Name</p>
+      "! Page Template Name
```
```suggestion
      "! Page Template name
```

ViktoriaFreidel replied:
we change all description. Thanks

wurzka commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,56 @@
+     general_information   TYPE ty_general_information,
+
+      "! <p class="shorttext">Page Templates</p>
+      "! Assigned SAP Fiori launchpad page templates
```
If lower case `page template` is fine here, I would go through all descriptions with `page template` and change the `P`s and `T`s to lower case

ViktoriaFreidel replied:
I will do it and check other description. Thanks for a Hint.

wurzka reviewed (commented):
I found some issues about the description. Could you please double check them, @ViktoriaFreidel ?
In contrast to the title, they shall not be in title case except when it is a common SAP term (like SAP **F**iori).

huber-nicolas commented on file-formats/uist/type/zif_aff_uist_v1.intf.abap:
```diff
@@ -0,0 +1,56 @@
+      name          TYPE c LENGTH 35,
+    END OF ty_page,
+
+    ty_pages TYPE TABLE OF ty_page WITH KEY name.
```
only a question: Do you want to have the entries sorted? If you want to have it as a standard table it would be more recognizable to explicitly declare it as standard table 

```suggestion
    ty_pages TYPE STANDARD TABLE OF ty_page WITH KEY name.
```

ViktoriaFreidel committed:
Update file-formats/uist/type/zif_aff_uist_v1.intf.abap

Co-authored-by: Michael Schneider <micha.schneider@sap.com>

ViktoriaFreidel committed:
Update file-formats/uist/examples/z_aff_example_uist.uist.json

Co-authored-by: Michael Schneider <micha.schneider@sap.com>

ViktoriaFreidel committed:
Update file-formats/uist/type/zif_aff_uist_v1.intf.abap

Co-authored-by: Michael Schneider <micha.schneider@sap.com>

ViktoriaFreidel committed:
Update file-formats/uist/type/zif_aff_uist_v1.intf.abap

Co-authored-by: Nicolas Huber <80829269+huber-nicolas@users.noreply.github.com>

ViktoriaFreidel committed:
[UIST] Add new object type UIST

ViktoriaFreidel committed:
[UIST] Add new object type UIST

ViktoriaFreidel committed:
[UIST] Add new object type UIST

schneidermic0 committed:
remove `multipleOf`from schema

schneidermic0 reviewed (approved):
Looks good for me, now.

We have to ignore the validation error for now. This is in progress to be fixed. See #498

ViktoriaFreidel commented:
@schneidermic0 : To the Question
"Did you also sign the contributor license agreement (CLA)?"

Answer: No, not yet. The Link to  CLA Page is not working at the moment.

DominikvonSchoenebeck commented:
Viktoria is out of office for a longer time.
Therefore I will act as substitute.

DominikvonSchoenebeck commented:
I could sign the CLA successfully.

schneidermic0 commented:
@DominikvonSchoenebeck The change is fine, now. You can consider it as reviewed

If you don't mind we won't merge this PR until the issue with JSON schema validation is fixed (see https://github.com/SAP/abap-file-formats/issues/498). The team is working on a fix and plans to provide it somewhen next week.

DominikvonSchoenebeck commented:
@schneidermic0 We are fine with that. Thanks for the review.

schneidermic0 committed:
Merge branch 'main' into feature/uist

schneidermic0 commented:
#498 is fixed. Thanks, @burcuka.

@ViktoriaFreidel @DominikvonSchoenebeck I'll merge this PR. Welcome as contributor 👍
